fix(ngcc): several fixes

gkalpak opened the pull request:
This PR includes several ngcc fixes. With these fixes, ngcc should be able to successfully process NativeScript. See individual commits for details.

Jira issue: [FW-1689][1]

##
TODO:

- [ ] Determine whether it is problematic that the synthetically created [function declaration][2] for TS helpers will be recognized as [non-synthetic][3] in `StaticInterpreter` (and how to fix it).
- [x] Implement the CommonJS fixes for UMD as well.
- [ ] Add integration tests.

[1]: https://angular-team.atlassian.net/browse/FW-1689
[2]: https://github.com/angular/angular/commit/e28ac708e#diff-3814db5d5b51f278653eb80c9ac981c1R235
[3]: https://github.com/angular/angular/blob/023c9bebe/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts#L242-L244



[FW-1689]: https://angular-team.atlassian.net/browse/FW-1689

mary-poppins commented:
You can preview 5fdc4a2 at https://pr34169-5fdc4a2.ngbuilds.io/.

mary-poppins commented:
You can preview 263efda at https://pr34169-263efda.ngbuilds.io/.

mary-poppins commented:
You can preview 8077ce1 at https://pr34169-8077ce1.ngbuilds.io/.

mary-poppins commented:
You can preview b0cc9d9 at https://pr34169-b0cc9d9.ngbuilds.io/.

mary-poppins commented:
You can preview a45d1bb at https://pr34169-a45d1bb.ngbuilds.io/.

mary-poppins commented:
You can preview 363218f at https://pr34169-363218f.ngbuilds.io/.

mary-poppins commented:
You can preview 08f17b8 at https://pr34169-08f17b8.ngbuilds.io/.

gkalpak committed:
fix(ngcc): correctly detect emitted TS helpers in ES5

In ES5 code, TypeScript requires certain helpers (such as
`__spreadArrays()`) to be able to support ES2015+ features. These
helpers can be either imported from `tslib` (by setting the
`importHelpers` TS compiler option to `true`) or emitted inline (by
setting the `importHelpers` and `noEmitHelpers` TS compiler options to
`false`, which is also the default value for both).

Ngtsc's `StaticInterpreter` (which is also used during ngcc processing)
is able to statically evaluate some of these helpers (currently
`__spread()` and `__spreadArrays()`), as long as
`ReflectionHost#getDefinitionOfFunction()` correctly detects the
declaration of the helper. For this to happen, the left-hand size of the
corresponding call expression (i.e. `__spread(...)` or
`tslib.__spread(...)`) must be evaluated as a function declaration for
`getDefinitionOfFunction()` to be called with.

In the case of imported helpers, the `tslib.__someHelper` expression was
resolved to a function declaration of the form
`export declare function __someHelper(...args: any[][]): any[];`, which
allows `getDefinitionOfFunction()` to correctly map it to a TS helper.

In contrast, in the case of emitted helpers (and regardless of the
module format: `CommonJS`, `ESNext`, `UMD`, etc.)), the `__someHelper`
identifier was resolved to a variable declaration of the form
`var __someHelper = (this && this.__someHelper) || function () { ... }`,
which upon further evaluation was categorized as a `DynamicValue`
(prohibiting further evaluation by the `getDefinitionOfFunction()`).

As a result of the above, TypeScript helpers were not evaluated in ES5
code.

---
This commit fixes this by detecting such helpers in
`Esm5ReflectionHost#getDeclarationOfIdentifier()` and (instead of the
actual variable declaration) return a synthesized function declaration,
identical to the one that would have been returned for the imported
helper.

This allows `getDefinitionOfFunction()` to correctly map it to a TS
helper, allowing `StaticInterpreter` to statically evaluate the
associated call expression.

gkalpak committed:
fix(ngcc): recognize more re-export formats in CommonJS

Previously, the `CommonJsReflectionHost` would only recognize re-exports
of the form `__export(require('...'))`. This is what re-exports look
like, when the TS helpers are emitted inline (i.e. when compiling with
the default [TS compiler options][1] that include `noEmitHelpers: false`
and `importHelpers: false`).

However, when compiling with `importHelpers: true` and [tslib][2] (which
is the recommended way for optimized bundles), the re-exports will look
like: `tslib_1.__exportStar(require('...'), exports)`
These types of re-exports where previously not recognized by
`CommonJsReflectionHost` and thus ignored.

This commit fixes this by ensuring both re-export formats are
recognized.

[1]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
[2]: https://www.npmjs.com/package/tslib

gkalpak committed:
fix(ngcc): correctly detect dependencies in CommonJS

Previously, `CommonJsDependencyHost.findDependencies()` would only find
dependencies via imports of the form `var foo = require('...');` or
`var foo = require('...'), bar = require('...');` However, CommonJS
files can have imports in many different forms. By failing to recognize
other forms of imports, the associated dependencies were missed, which
in turn resulted in entry-points being compiled out-of-order and failing
due to that.

While we cannot easily capture all different types of imports, this
commit enhances `CommonJsDependencyHost` to recognize the following
common forms of imports:

- Imports in property assignments. E.g.:
  `exports.foo = require('...');` or
  `module.exports = {foo: require('...')};`

- Imports for side-effects only. E.g.:
  `require('...');`

- Star re-exports (with both emitted and imported heleprs). E.g.:
  `__export(require('...'));` or
  `tslib_1.__exportStar(require('...'), exports);`

mary-poppins commented:
You can preview e28ac70 at https://pr34169-e28ac70.ngbuilds.io/.

petebacondarwin commented:
typo in the 2nd to last commit: `findDepeendencies`

petebacondarwin commented:
typos in the last commit:

> In the case of emitted helpers, the `tslib.__someHelper` expression was
resolved to a function declaration of the form
`export declare function __someHelper(...args: any[][]): any[];`, which
allows `getDefinitionOfFunction()` to correctly map it to a TS helper.

I think this should be `In the case of imported helpers, ...`.

And also `functinos`

petebacondarwin commented:
typo in 5th commit: By failing to recongize

petebacondarwin commented on packages/compiler-cli/ngcc/test/dependencies/commonjs_dependency_host_spec.ts:
```diff
@@ -12,6 +12,29 @@ import {loadTestFiles} from '../../../test/helpers';
+  // tslib_1.__exportStar(require('...'), exports);
+  reExportsWithImportedHelper?: string[];
+}
+
```
NIT: I prefer not to clutter the start of a test file with supporting stuff (like helper functions and interfaces) since it makes it more difficult to quickly get to the tests when browsing later. Since interfaces are always hoisted, can we move this down below the `describe` block?

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/commonjs_dependency_host.ts:
```diff
@@ -40,33 +40,66 @@ export class CommonJsDependencyHost extends DependencyHostBase {
+                            ts.isPropertyAssignment(prop) && isRequireCall(prop.initializer))
+                    .map(prop => prop.initializer);
+
+            requireCalls.push(...requireCallsFromProperties);
```
NIT: this could be more simple and not require intermediate arrays to be created:

```ts
        stmt.expression.right.properties.forEach(prop => {
          if(ts.isPropertyAssignment(prop) && isRequireCall(prop.initializer)) {
            requireCalls.push(prop.initializer);
          }
        });
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/commonjs_dependency_host.ts:
```diff
@@ -40,33 +40,66 @@ export class CommonJsDependencyHost extends DependencyHostBase {
+      }
+    }
+
+    const importPaths = new Set(requireCalls.map(call => call.arguments[0].text));
```
Nice!

petebacondarwin reviewed (commented):
Reviewed all but the last comment.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -209,6 +218,40 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
+            ],
+            ts.createArrayTypeNode(ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)), undefined),
+        viaModule: null,
+      };
```
NIT: I would put this whole block into its own function (e..g `createSynthesizedHelperDeclaration(id)`) with the  big comment converted to JSDOC comment. This will keep the calling function much smaller and more readable.

petebacondarwin reviewed (approved):
The PR looks great! Regarding the last commit, I can accept this approach. I don't know what other ideas you had. One thought is if we could create a new thing called a `LazyFunction`, which we would match code of the  form `(this && this.<fn>) || function () { ... }`, which could then be visited more intelligently in the StaticInterpreter?

petebacondarwin commented:
@gkalpak is this PR nearly ready? Is it worth splitting out the non-contentious bits so that we can land them this week?

gkalpak commented:
I'll split the non-contentious bits into a separate PR this week 👍

gkalpak committed:
fixup! fix(ngcc): recognize more re-export formats in CommonJS

gkalpak committed:
fixup! fix(ngcc): correctly detect dependencies in CommonJS

gkalpak committed:
fixup! fix(ngcc): correctly detect emitted TS helpers in ES5

gkalpak commented:
@petebacondarwin: I have addressed your comments (in `fixup` commits - except for the commit message typos) and moved some of the commits into separate PRs: #34437, #34441

I need to further split this into more PRs (for easier reviewing) and add similar fixes for UMD (where appropriate).

mary-poppins commented:
You can preview 8c37785 at https://pr34169-8c37785.ngbuilds.io/.

gkalpak commented:
Further split this up into separate PRs per fix: #34512, #34527, #34528
<sub>(Each builds on top of the previous one, so you can either only review all commits on the last one or only review the new commits on each one.)</sub>

gkalpak commented:
With all the "spin-off" PRs merged, the only remaining fix for this PR is to **correctly detect emitted TS helpers in ES5** (i.e. 0b89f17c3). (I need to clean it up or possibly rework the approach.)

This will also fix angular/ngcc-validation#762.

gkalpak commented:
The final fix is in #35139. This PR is now obsolete. Closing...
